pkg/controller/storage/capability/capability_controller.go

```diff
@@ -0,0 +1,317 @@
+	storageClassCapabilityClient clientset.Interface
+	storageClassLister           sclisters.StorageClassLister
+	storageClassSynced           cache.InformerSynced
+	snapshotClassLister          snaplisters.VolumeSnapshotClassLister
+	snapshotClassSynced          cache.InformerSynced
+	storageClassCapabilityLister crdlisters.StorageClassCapabilityLister
+	storageClassCapabilitySynced cache.InformerSynced
+	workQueue                    workqueue.RateLimitingInterface
+	csiAddressGetter             csiAddressGetter
+}
+
+// This controller is responsible to watch StorageClass, SnapshotClass.
+// And then update StorageClassCapability CRD resource object to the newest status.
+func NewController(
+	k8sClient kubernetes.Interface,
+	storageClassCapabilityClient clientset.Interface,
+	storageClassInformer scinformers.StorageClassInformer,
+	snapshotClassInformer snapinformers.VolumeSnapshotClassInformer,
+	storageClassCapabilityInformer storageinformers.StorageClassCapabilityInformer,
+	csiAddressGetter csiAddressGetter,
+) *StorageCapabilityController {
+
+	utilruntime.Must(crdscheme.AddToScheme(scheme.Scheme))
+	controller := &StorageCapabilityController{
+		k8sClient:                    k8sClient,
+		storageClassCapabilityClient: storageClassCapabilityClient,
+		storageClassLister:           storageClassInformer.Lister(),
+		storageClassSynced:           storageClassInformer.Informer().HasSynced,
+		snapshotClassLister:          snapshotClassInformer.Lister(),
+		snapshotClassSynced:          snapshotClassInformer.Informer().HasSynced,
+		storageClassCapabilityLister: storageClassCapabilityInformer.Lister(),
+		storageClassCapabilitySynced: storageClassCapabilityInformer.Informer().HasSynced,
+		workQueue:                    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "StorageClasses"),
+		csiAddressGetter:             csiAddressGetter,
+	}
+	storageClassInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueueStorageClass,
+		UpdateFunc: func(old, new interface{}) {
+			newStorageClass := new.(*v1strorage.StorageClass)
+			oldStorageClass := old.(*v1strorage.StorageClass)
+			if newStorageClass.ResourceVersion == oldStorageClass.ResourceVersion {
+				return
+			}
+			controller.enqueueStorageClass(newStorageClass)
+		},
+		DeleteFunc: controller.enqueueStorageClass,
+	})
+	snapshotClassInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: controller.enqueueSnapshotClass,
+		UpdateFunc: func(old, new interface{}) {
+			return
+		},
+		DeleteFunc: controller.enqueueSnapshotClass,
+	})
+	return controller
+}
+
+func (c *StorageCapabilityController) Start(stopCh <-chan struct{}) error {
+	return c.Run(5, stopCh)
+}
+
+func (c *StorageCapabilityController) Run(threadCnt int, stopCh <-chan struct{}) error {
+	defer utilruntime.HandleCrash()
+	defer c.workQueue.ShutDown()
+
+	isValid, err := c.IsValidKubernetesVersion()
```

Comment:
I think we should move this version check before controller initialization, it's not efficient after controller is running.

Reply:
OK, I will fix it.